### ISSUE TITLE
duplicate documents getting generated on save

### ISSUE THREAD

**Issue by codenomnom** [sdorra/content-collections#373]
Hi there 👋 Thanks for the amazing library - it saves my day! 🥳 

I'm having this weird issue I can't understand how to debug properly. I am using Next.js with the appropriate setup. I have a collection and it works like a charm when I start the next app for the first time. I see my collection of files and they render on the front page properly.

But then I go and just _save_ one of my files again. Regardless if I make changes or not. Then I have my file duplicated inside the collection. Here's the output (which is also weird):

```bash
# here I've hit ctrl+s inside my content-collections.ts
... config changed
build started ...
... finished build of 1 collection and 9 documents in 6ms
# all good so far

# here I hit ctrl+s inside file1.json, without making any changes
... file src\content\file1.json changed
build started ...
... finished build of 1 collection and 10 documents in 5ms
# somehow documents got 10, even though I haven't created another file

# weirdly enough, if I hit ctrl+s again (second time), two events get fired:
... file src\content\file1.json changed
build started ...
... finished build of 1 collection and 10 documents in 5ms
... file src\content\file1.json changed
build started ...
... finished build of 1 collection and 10 documents in 5ms

# hitting ctrl+s more times results in just two events getting fired..

# but if I ctrl+save my content-collections.ts again, I see this:
... config changed
build started ...
... finished build of 1 collection and 9 documents in 7ms
```

I'm seeing this on Windows 11, TS 5.6.3, with this following project setup:
```json
"devDependencies": {
  "@content-collections/core": "^0.7.2",
  "@content-collections/next": "^0.2.3",
  "@types/node": "^20",
  "@types/react": "^18",
  "@types/react-dom": "^18",
  "eslint": "^8",
  "eslint-config-next": "15.0.1",
  "postcss": "^8",
  "tailwindcss": "^3.4.1",
  "typescript": "^5"
}
```

Any ideas on how to troubleshoot it further? I'll probably clone the repo tomorrow and play with it, but I'd appreciate any hints upfront 😉

**Comment by sdorra** [comment]
I assume that the path of the document is not created correctly in the [synchronizer](https://github.com/sdorra/content-collections/blob/main/packages/core/src/synchronizer.ts#L36), so that content collections think this is a new document instead of an update. It looks like it happens only on Windows, so perhaps it has something to do with the directory separator.

**Comment by codenomnom** [comment]
That's what I thought. Unfortunately I couldn't dig much deeper. I cloned the project and tried running the tests. The majority of them are failing, mostly because of this:

```
X [ERROR] The entry point "C:\\content-collections\\packages\\core\\src\\__tests__\\config.001.ts" cannot be marked as external
```

It leads to `esbuild.ts::56` where it's calling the `build` function. The path there is correct: `C:\content-collections\packages\core\src\__tests__\config.001.ts`, but it seems it cannot make it work. Running the `dev` tasks fails with the same reason 😞 

---

Apart from that I am able to build the project and use some sample setup like creating a new collection and printing its internals. Not of much help though.

The only thing that comes to my mind is create a new empty project, building the source code, copying it inside the project and initializing it using the import from the new `dist` I've created. Then trying the watcher.
Any ideas on how to test/debug it better?

**Comment by codenomnom** [comment]
Okay, something meaningful 😁 I figured I can loop through the collection manually and see what's inside, including the meta. Surely enough, it's Windows 😉 

```ts
collection: [ // mapped item to item._meta.filePath
  'content/file1.json',
  'content\\file1.json',
]
```

So I guess the problem is that whenever the code is run the first time, it manually collects and builds the collection. Then, the file watcher sends an update to a "new file", so it's added again.
And I'm also guessing the easiest way to fix the issue is inside `builder.ts::83 (sync function)` by overriding the `filePath` with unix separator. Otherwise the rest of the code uses built-in `node:path`, which do not cause any issue and too many changes would be required.

I'm unsure if this is the right place to fix, nor if it will fix it (haven't boot up my Ubuntu machine in a while), but if needed - I can open a PR for that so that you can test it 😉 Just let me know. Thanks!

**Comment by codenomnom** [comment]
Haha I've pasted the `devDependencies` 😁I'm using Next `15.0.1`, here are a few code samples:

`next.config.ts`
```ts
import type { NextConfig } from "next";
import { withContentCollections } from "@content-collections/next";

const nextConfig: NextConfig = {};
export default withContentCollections(nextConfig);
```

`tsconfig.json`
```json
{
  "compilerOptions": {
    "target": "ES2017",
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "bundler",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "incremental": true,
    "plugins": [
      {
        "name": "next"
      }
    ],
    "paths": {
      "@/*": ["./src/*"],
      "content-collections": ["./.content-collections/generated"]
    }
  },
  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx", ".next/types/**/*.ts"],
  "exclude": ["node_modules"]
}
```

`content-collections.ts`
```ts
import { defineCollection, defineConfig } from '@content-collections/core';
 
const ui = defineCollection({
  name: 'ui',
  directory: 'src/content',
  include: '**/*.json',
  schema: (z) => ({}), // skipping actual schema
  parser: 'json',
});
 
export default defineConfig({
  collections: [ui],
});
```

I think that's all I got. Basically a stock Next.js project with content-collections on top. I'm currently using VSCode on this machine - that might be another point of failure as it's a shady one🤣

**Comment by sdorra** [comment]
> If you can't reproduce it I'll upload my repo online so you can check it.

Yes, please upload it. I've tried it with the latest next.js, but it works as expected:

```
build started ...
... finished build of 1 collection and 1 document in 23ms
start watching ...
  ▲ Next.js 15.0.1 (Turbopack)
  - Local:        http://localhost:3000

 ✓ Starting...
 ✓ Ready in 2.5s
... file src\content\one.json changed
build started ...
... finished build of 1 collection and 1 document in 6ms
... file src\content\one.json changed
build started ...
... finished build of 1 collection and 1 document in 8ms
... file src\content\one.json changed
build started ...
... finished build of 1 collection and 1 document in 13ms
... file src\content\one.json changed
build started ...
... finished build of 1 collection and 1 document in 20ms
```

**Comment by sdorra** [comment]
I was also able to fix all unit tests from core on Windows, except for 2.

**Comment by codenomnom** [comment]
This one got me good 😺 I created a new project and initially I couldn't reproduce 😨 It turns out the issue only occurs when the files are in nested folders:

```ts
const posts = defineCollection({
  name: "posts",
  directory: "src/content",
  include: "**/*.json",
  schema: (z) => ({
    title: z.string(),
  }),
  parser: "json",
});
```

And then the files:
```
src
└── content
    ├── sub-folder
    │   └── file1.json <-- this one IS causing issues
    └── root.json <-- this one NOT causing issues
```

I've uploaded the sample repo here: https://stackblitz.com/edit/content-collections-duplication-example Keep in mind it does work _correctly_ inside  the online editor (since it's not Windows 😁), but the same repo would fail on my machine.

I've also recorded a video of what's happening - it's only a terminal trace, but you can see collection number increases. I've first hit `ctrl+s` in `content-collections.ts` file, then done the same in `file3.json` and `file2.json`, which is nested and thus causing issues. I hope that helps!


https://github.com/user-attachments/assets/fa2570f6-b1aa-4a67-97e4-9c23a1fd56be

**Comment by codenomnom** [comment]
I can also confirm tests are running okay now, except the `writer :: os specific` ones:

```
 FAIL  src/writer.test.ts > writer > os specific > should write import with / instead of \ on windows
AssertionError: expected 'import configuration from "../C:\User…' to contain 'import configuration from "./sub/conf…'

- Expected
+ Received

- import configuration from "./sub/config.ts";
+ import configuration from "../C:\Users\cnn\AppData\Local\Temp\vitest-XXXXXXadCcMt\sub\config.ts";
+ import { GetTypeByName } from "@content-collections/core";
+
+ export type Test = GetTypeByName<typeof configuration, "test">;
+ export declare const allTests: Array<Test>;
+
+ export {};
+
```

**Comment by codenomnom** [comment]
One more thing - the watcher seems to use the native separators for the OS (https://github.com/parcel-bundler/watcher/issues/50) so I guess it's inevitable to "fix" them internally.

**Comment by sdorra** [comment]
Ok, i've done some more tests and i could reproduce the issue. I can also confirm that the initial collected files use `/` for the path and the synced ones use `\`. But replacing `path.win32.sep` with `path.posix.sep` on the [watcher](https://github.com/sdorra/content-collections/blob/7a21995a45dd34a6b24c9740042389aa8fe53308/packages/core/src/watcher.ts#L47) does not fix the problem. The `\` comes back after [synchronizer#createRelativePath](https://github.com/sdorra/content-collections/blob/7a21995a45dd34a6b24c9740042389aa8fe53308/packages/core/src/synchronizer.ts#L25), if i do the replace before returning the [relativePath](https://github.com/sdorra/content-collections/blob/7a21995a45dd34a6b24c9740042389aa8fe53308/packages/core/src/synchronizer.ts#L33C12-L33C24) the problem is fixed.

My problem is that i don't understand why the `\` comes back. I will do some research and some more tests.

**Comment by sdorra** [comment]
After researching and after a few more tests, i've come to the following conclusion: Node.js uses `\` as the separator in all its functions (e.g.: `path.join`, `path.resolve`, `path.relative`) and this sounds absolute logically to me. But why we have paths with `/` in our collection ... it turns out that the globbing library ([tinyglobby](https://github.com/SuperchupuDev/tinyglobby)) uses [posix path regardless of the platform it is running on](https://github.com/SuperchupuDev/tinyglobby/blob/b9aa5fd0f151330fe7aa96a2b419689fd9346665/src/index.ts#L50).

So i think the right way to fix this issue, is to replace the posix separator (`/`) with the windows separator (`\`) after the globbing in the [collector](https://github.com/sdorra/content-collections/blob/7a21995a45dd34a6b24c9740042389aa8fe53308/packages/core/src/collector.ts#L101).

**Comment by codenomnom** [comment]
That's a lovely story! 😄 I personally would pick posix (`/`) as `\` needs to be (un)escaped everywhere it's used, but it's up to you to figure that out 😉 I'm glad you could find the problem.

**Comment by sdorra** [comment]
The fix is now released with version 0.7.3 of `@content-collections/core`
